pkg/sources/udsource/grpc_udsource.go

```diff
@@ -0,0 +1,161 @@
+/*
+Copyright 2022 The Numaproj Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package udsource
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	sourcepb "github.com/numaproj/numaflow-go/pkg/apis/proto/source/v1"
+
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	"github.com/numaproj/numaflow/pkg/isb"
+	sourceclient "github.com/numaproj/numaflow/pkg/sdkclient/source/client"
+	"github.com/numaproj/numaflow/pkg/sources/udsource/utils"
+)
+
+// GRPCBasedUDSource applies a user-defined source over gRPC
+// connection where server is the UDSource.
+type GRPCBasedUDSource struct {
+	client sourceclient.Client
+}
+
+// NewUDSgRPCBasedUDSource accepts a gRPC client and returns a new GRPCBasedUDSource.
+func NewUDSgRPCBasedUDSource(c sourceclient.Client) (*GRPCBasedUDSource, error) {
+	return &GRPCBasedUDSource{c}, nil
+}
+
+// CloseConn closes the gRPC client connection.
+func (u *GRPCBasedUDSource) CloseConn(ctx context.Context) error {
+	return u.client.CloseConn(ctx)
+}
+
+// IsHealthy checks if the udsource is healthy.
+func (u *GRPCBasedUDSource) IsHealthy(ctx context.Context) error {
+	return u.WaitUntilReady(ctx)
+}
+
+// WaitUntilReady waits until the udsource is connected.
+func (u *GRPCBasedUDSource) WaitUntilReady(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("failed on readiness check: %w", ctx.Err())
+		default:
+			if _, err := u.client.IsReady(ctx, &emptypb.Empty{}); err == nil {
+				return nil
+			}
+			time.Sleep(1 * time.Second)
```

Comment:
I am wondering whether we should log the error and sleep so users will know what is happening, esp. if someone has attached the wrong container to the vertex. other wise, it is an infinite wait.